src/cmd_all/src/single_node.rs

```diff
@@ -227,3 +268,256 @@ impl SingleNodeOpts {
+        let _compute_handle = tokio::spawn(async move { risingwave_compute::start(opts).await });
+    }
+    if let Some(opts) = frontend_opts {
+        tracing::info!(?opts, "starting frontend-node service");
+        let _frontend_handle = tokio::spawn(async move { risingwave_frontend::start(opts).await });
+    }
+    if let Some(opts) = compactor_opts {
+        tracing::info!(?opts, "starting compactor-node service");
+        let _compactor_handle =
+            tokio::spawn(async move { risingwave_compactor::start(opts).await });
+    }
+
+    // wait for log messages to be flushed
+    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+    eprintln!("-------------------------------");
+    eprintln!("RisingWave single_node mode is ready.");
+
+    // TODO: should we join all handles?
+    // Currently, not all services can be shutdown gracefully, just quit on Ctrl-C now.
+    // TODO(kwannoel): Why can't be shutdown gracefully? Is it that the service just does not
+    // support it?
+    signal::ctrl_c().await.unwrap();
+    tracing::info!("Ctrl+C received, now exiting");
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use std::fmt::Debug;
+
+    use expect_test::{expect, Expect};
+
+    use super::*;
+
+    fn check(actual: impl Debug, expect: Expect) {
+        let actual = format!("{:#?}", actual);
+        expect.assert_eq(&actual);
+    }
+
+    #[test]
+    fn test_parse_extra_opts() {
+        let raw_opts = "
+--meta-extra-opts=--advertise-addr 127.0.0.1:9999 --data-directory \"some path with spaces\" \
+--state-store hummock+fs:///root/.risingwave/state_store \
+--sql-endpoint sqlite:///root/.risingwave/meta_store/single_node.db?mode=rwc
+--compute-extra-opts=--listen-addr 127.0.0.1:8888 --total-memory-bytes 123 --parallelism 10
+";
+        let actual = SingleNodeOpts::parse_from(raw_opts.lines());
+        check(
+            &actual,
+            expect![[r#"
+                SingleNodeOpts {
+                    prometheus_listener_addr: None,
+                    config_path: None,
+                    store_directory: None,
+                    meta_addr: None,
+                    compute_addr: None,
+                    frontend_addr: None,
+                    compactor_addr: None,
+                    meta_extra_opts: Some(
+                        "--advertise-addr 127.0.0.1:9999 --data-directory \"some path with spaces\" --state-store hummock+fs:///root/.risingwave/state_store --sql-endpoint sqlite:///root/.risingwave/meta_store/single_node.db?mode=rwc",
+                    ),
+                    compute_extra_opts: Some(
+                        "--listen-addr 127.0.0.1:8888 --total-memory-bytes 123 --parallelism 10",
+                    ),
+                    frontend_extra_opts: None,
+                    compactor_extra_opts: None,
+                }"#]],
+        );
+        let actual_parsed = parse_single_node_opts(&actual).unwrap();
```

Comment:
nit: this looks a little strange (we first "parse", and then "parse" again..? 🤣):
```rust
let actual = SingleNodeOpts::parse_from(raw_opts.lines());
let actual_parsed = parse_single_node_opts(&actual).unwrap();
```
Maybe `build` or `convert` or sth

Reply:
Changing it to `normalize`. Since we normalize the opts from various levels into node level options.